packages/mobile/src/web3/contracts.ts

```diff
@@ -1,128 +1,112 @@
+/**
+ * This file is called 'contracts' but it's responsibilies have changed over time.
+ * It now manages contractKit and gethWallet initialization.
+ * Leaving the name for recognizability to current devs
+ */
 import { ContractKit, newKitFromWeb3 } from '@celo/contractkit'
 import { RpcWallet } from '@celo/contractkit/lib/wallets/rpc-wallet'
-import { Platform } from 'react-native'
-import * as net from 'react-native-tcp'
-import { call, select, take } from 'redux-saga/effects'
+import { sleep } from '@celo/utils/src/async'
+import { call, select } from 'redux-saga/effects'
 import { DEFAULT_FORNO_URL } from 'src/config'
-import { IPC_PATH } from 'src/geth/geth'
-import { waitForGethConnectivity } from 'src/geth/saga'
-import { promisifyGenerator } from 'src/utils/generators'
+import { waitForGethInitialized } from 'src/geth/saga'
 import Logger from 'src/utils/Logger'
-import { Actions } from 'src/web3/actions'
-import { contractKitReadySelector, fornoSelector } from 'src/web3/selectors'
+import { getHttpProvider, getIpcProvider } from 'src/web3/providers'
+import { fornoSelector } from 'src/web3/selectors'
 import Web3 from 'web3'
-import { IpcProvider, provider } from 'web3-core'
+import { IpcProvider } from 'web3-core'
 
-const tag = 'web3/contracts'
+const TAG = 'web3/contracts'
 
 let ipcProvider: IpcProvider | undefined
 let gethWallet: RpcWallet | undefined
 let contractKit: ContractKit | undefined
 
-export function openIpcProvider() {
-  if (!ipcProvider) {
-    Logger.debug(`${tag}@openIpcProvider`, `Initializing IPC connection`)
-    ipcProvider = getIpcProvider()
-  }
-  return ipcProvider
-}
+export function* initContractKit() {
+  // The kit must wait for Geth to be initialized because
+  // Geth is required for the RpcWallet
+  yield call(waitForGethInitialized)
 
-export function openWallet() {
-  if (!gethWallet) {
-    Logger.debug(`${tag}@openWallet`, `Initializing wallet`)
-    gethWallet = new RpcWallet(openIpcProvider())
+  if (contractKit || ipcProvider || gethWallet) {
+    throw new Error('Kit not properly destroyed')
   }
-  return gethWallet
-}
 
-export function openContractKit(fornoMode: boolean) {
-  if (!contractKit) {
-    Logger.info(
-      `${tag}@openContractKit`,
-      `Initializing contractkit, platform: ${Platform.OS}, forno mode: ${fornoMode}`
-    )
-    const web3 = new Web3(fornoMode ? getHttpProvider(DEFAULT_FORNO_URL) : openIpcProvider())
-    const wallet = openWallet()
-    contractKit = newKitFromWeb3(web3, wallet)
-  }
-  return contractKit
+  const fornoMode = yield select(fornoSelector)
+  Logger.info(`${TAG}@initContractKit`, `Initializing contractkit, forno mode: ${fornoMode}`)
+
+  ipcProvider = getIpcProvider()
+  const web3 = new Web3(fornoMode ? getHttpProvider(DEFAULT_FORNO_URL) : ipcProvider)
+
+  Logger.debug(`${TAG}@initContractKit`, `Initializing wallet`)
+  gethWallet = new RpcWallet(ipcProvider)
```

Comment:
do we need a gethWallet if we aren't in forno mode? cc @yorhodes

Reply:
Is all private key storage handled in the RPC wallet? It could be worth keeping if we decide to switch to clef down the line

Reply:
Is there a downside to using it? Splitting the code paths would mean more code to test.

Reply:
I guess just memory as well as potentially more round trips over the RPC (local)